src/boot/efi/efi-string.c

```diff
@@ -173,6 +173,86 @@ char16_t *xstrdup16(const char16_t *s) {
         return xstrndup16(s, SIZE_MAX);
 }
 
+bool metaimatch(const char16_t *haystack, const char16_t *pattern) {
+        assert(pattern);
+        assert(haystack);
+
+        for (;;) {
+                char16_t h = TOLOWER(*haystack), p = TOLOWER(*pattern);
+
+                switch (p) {
+                case '\0':
+                        /* End of pattern. Check that haystack is now empty. */
+                        return h == '\0';
+
+                case '?':
+                        if (h == '\0')
+                                /* Premature end of haystack. */
+                                return false;
+                        break;
+
+                case '*':
+                        do {
+                                /* Try matching haystack with remaining pattern. */
+                                if (metaimatch(haystack, pattern + 1))
+                                        return true;
+
+                                /* Otherwise, we match one char here. */
+                                haystack++;
+                        } while (*haystack);
+
+                        /* End of haystack. Pattern needs to be empty too for a match. */
+                        return *(pattern + 1) == '\0';
+
+                case '[':
+                        if (h == '\0')
+                                /* Premature end of haystack. */
+                                return false;
+
+                        for (char16_t low = 0;;) {
+                                pattern++;
+                                p = TOLOWER(*pattern);
+                                if (!p || p == ']')
```

Comment:
`metaimatch` is intentionally written to exhibit the same (quirky) behavior as EDK2 `MetaiMatch` (since most firmware is written against that and we eventually want to call the firmware implementation). This probably wasn't clear enough, so I added comments to clarify. The previous version wasn't 100% there, which is now fixed. I even made sure by verifying all the test cases against the EDK2 function. I also adjusted the man page to hopefully make it clear it's not 100% fnmatch.